### PR TITLE
Save the 'be consistent' text

### PR DESCRIPTION
**Pull request by martinthomson**
This rescues the good text from #25, on which we are converging on consensus to reject.

### CODE REVIEW

File: draft-thomson-http-replay.md [hunk 1]
```diff
@@ -287,6 +289,37 @@ forwarding the request until the TLS handshake completes, or send a 4NN (Too
...
 
+## Consistent Handling of Early Data {#be-consistent}
+
+Consistent treatment of a request that arrives in - or partially in - early data
+is critical to avoiding inappropriate processing of replayed requests.  Requests
+that might be processed prior to the handshake completing MUST be processed the
+same way, even if the handshake is complete when processing occurs.
+
+A TLS implementation MUST ensure that it is possible to correctly identify early
+data correctly.  This is necessary even if early data arrives at the same time
```
Comment (kazuho) [inline draft-thomson-http-replay.md]: superfluous "correctly"?

Comment (martinthomson) [reply draft-thomson-http-replay.md]: thanks

**Comment by wtarreau** [comment]
All of this looks perfect to me, thanks for doing it Martin. Be careful on 2nd patch however, you have a typo ("servers can initiated").

File: draft-thomson-http-replay.md [hunk 2]
```diff
@@ -287,6 +289,37 @@ forwarding the request until the TLS handshake completes, or send a 4NN (Too
 Early) status code in response.  A gateway that is uncertain about whether an
 origin server supports the `Early-Data` header field SHOULD disable early data.
 
+## Consistent Handling of Early Data {#be-consistent}
+
+Consistent treatment of a request that arrives in - or partially in - early data
+is critical to avoiding inappropriate processing of replayed requests.  Requests
+that might be processed prior to the handshake completing MUST be processed the
+same way, even if the handshake is complete when processing occurs.
```
Comment (mnot) [inline draft-thomson-http-replay.md]: The same way as what?

File: draft-thomson-http-replay.md [hunk 3]
```diff
@@ -287,6 +289,37 @@ forwarding the request until the TLS handshake completes, or send a 4NN (Too
...
+
+A TLS implementation MUST ensure that it is possible to correctly identify early
```
Comment (mnot) [inline draft-thomson-http-replay.md]: On the face of it, this seems like it's putting requirements upon TLS implementations. A better way might be to say "To be used for early data in HTTP, a TLS implementation MUST..."

File: draft-thomson-http-replay.md [hunk 4]
```diff
@@ -287,6 +289,37 @@ forwarding the request until the TLS handshake completes, or send a 4NN (Too
...
+data.  This is necessary even if early data arrives at the same time as the TLS
+Finished message from the client.  If early data is misclassified, requests
+might erroneously appear to be safe to handle.  Even though the handshake might
+complete successfully on this connection, this is no guarantee that the same
+early data was not processed prior to handshake completion on another
+connection where the early data was replayed.
```
Comment (mnot) [inline draft-thomson-http-replay.md]: Is this last sentence dependent upon the previous ones, or is it a separate point that deserves its own paragraph? It seems to me that this is an important bit of guidance to give to servers when they're deciding whether to accept early data, but it's buried in a paragraph that starts with requirements about TLS implementations.

File: draft-thomson-http-replay.md [hunk 5]
```diff
@@ -287,6 +289,37 @@ forwarding the request until the TLS handshake completes, or send a 4NN (Too
...
+
+A server MUST NOT process requests that arrive in early data, or requests marked
+with the `Early-Data` header field unless it determines that the processing is
+free from side-effects.  Otherwise, the server MUST consistently use the same
+approach: either using the 4NN status code, or delaying processing until the
+handshake completes.
+
+A server MUST use the 4NN status code if the request is marked with the
+Early-Data header field.  Presence of the header field indicates that early data
```
Comment (mnot) [inline draft-thomson-http-replay.md]: This reads like a requirement to ALWAYS respond to Early-Data with 4NN.

File: draft-thomson-http-replay.md [hunk 6]
```diff
@@ -287,6 +289,37 @@ forwarding the request until the TLS handshake completes, or send a 4NN (Too
...
+was used on a previous hop.  Completion of the TLS handshake on that hop isn't
+visible to the server.
+
+Servers can sometimes initiate processing on requests that are only partially
+received.  These considerations apply to any processing that the server might
```
Comment (mnot) [inline draft-thomson-http-replay.md]: Which considerations?

File: draft-thomson-http-replay.md [hunk 7]
```diff
@@ -287,6 +289,37 @@ forwarding the request until the TLS handshake completes, or send a 4NN (Too
...
+apply.  Parsing header fields and request routing is likely to be safe from
```
Comment (mnot) [inline draft-thomson-http-replay.md]: Probably be explicit; e.g., "Parsing header fields -- but not acting upon them -- and ..."

**Comment by martinthomson** [comment]
I trimmed this right down.  I'm not sure if it is needed now.

**Comment by martinthomson** [comment]
After more list discussion, I think that we should remove the TLS implementation requirements, and add a small piece on consistency.  I will open a separate PR on the operational considerations.

**Comment by wtarreau** [comment]
At least we need consistency on the TLS behaviour. In this case we can
be fine with TLS implementations not telling the difference between
*validated* early data and regular data.

**Comment by martinthomson** [comment]
I think that we already have the "is the handshake complete" signal.  Do you think that we need explicit text here?  I worry that levying a requirement on that signal is a little ridiculous.

**Comment by wtarreau** [comment]
If we have this signal all the time (if you remember we were having a doubt for NSS prior to your verification) it'd be better not to add text that could possibly confuse implementers. I'm fine with the proposal as it is. Thanks.

**Pull request merged**